### DISCUSSION TITLE
Mixing version constraints with git dependencies

### DISCUSSION THREAD

**Discussion by dolfim-ibm** [python-poetry/poetry#7447] [Q&A]
We have a tree of dependencies, all being defined as `git` dependencies,  which looks like the following
```
repo1
  - pkg_a
  - pkg_i
    - pkg_a
  - pkg_j
    - pkg_a
  - pkg_k
    - pkg_a
```

**repo1/pyproject.toml**
```
[tool.poetry.dependencies]
pkg_a = {git = "ssh://git@mygit.com/pkg_a.git", branch = "new-feature"}
pkg_i = {git = "ssh://git@mygit.com/pkg_i.git"}
pkg_j = {git = "ssh://git@mygit.com/pkg_j.git"}
pkg_k = {git = "ssh://git@mygit.com/pkg_k.git"}
```

**pkg_i/pyproject.toml**
```
[tool.poetry.dependencies]
pkg_a = {git = "ssh://git@mygit.com/pkg_a.git"}
```

**pkg_j/pyproject.toml**
```
[tool.poetry.dependencies]
pkg_a = {git = "ssh://git@mygit.com/pkg_a.git"}
```

**pkg_k/pyproject.toml**
```
[tool.poetry.dependencies]
pkg_a = {git = "ssh://git@mygit.com/pkg_a.git"}
```

### Our problem

When developing a new feature in the shared packaged `pkg_a`, we cannot "simply" update the `repo1/pyproject.toml`, because the other sub-packages are still pointing to the `main` branch.
We get error resolutions like
```
Because pkg_i (0.1.0) @ git+ssh://git@mygit.com/pkg_i.git depends on pkg_a (0.4.0) @ git+ssh://git@mygit.com/pkg_a.git
 and repo depends on pkg_a (0.4.1) @ git+ssh://git@mygit.com/pkg_a.git@new-feature, pkg_i is forbidden.
So, because repo depends on pkg_i (0.1.0) @ git+ssh://git@mygit.com/pkg_i.git, version solving failed.

```

Currently, we seems to be forced in re-pinning the feature branch in all `pkg_i`, `pkg_j`, `pkg_k`, which is not very convenient.

We were wondering if there is already another way of doing it.
Maybe something like the common usage of version constraints?
What if the `main` branch of `pkg_a` is set to version `0.4.0` and the `new-feature` would be `0.4.1`? Shouldn't poetry be able to resolve that, similar to setting `version=^0.4.0`?

**Comment by neersighted** [comment]
Git tags/branches have no special semantics. In this situation, where you have multiple projects with shared dependencies developed independently, you really do need a package index.
